### ISSUE TITLE
Off-center sound effects

### ISSUE THREAD

**Issue by geist22** [swfans/swars#134]
Does Syndicate Wars use stereo sound? I didn't think it did...

Playing through DOSBox, all the sounds are centered. But in the port, I feel like I'm hearing some sound effects more from the left speaker than the right. It's particularly noticeable with the idle speech from the policemen, but also the alarm from High Explosives or the sound of fire are noticeable.

I've checked the oldest release I've got working, 0.3.1.107, and it already has this issue.

**Comment by Moburma** [comment]
Yeah, I've noticed this as well, it's hard to tell where a sound is coming from, it's often reversed so that e.g. you're hearing a sound to the left that is actually happening off screen to the right.

**Comment by mefistotelis** [comment]
I don't think "having 3D sound" is an issue by itself.

If the 3D sound doesn't correctly compute the direction, that's a different matter.
Can this be tested on something which emits all the time, like a tank?

This was part of the original Miles Sound System, but OpenAL implementation is completely different.

\<offtopic\>

Actually, OpenAL supports even doppler effects from movement. Though that part isn't used.

I was thinking about makin proper doppler-based sound for cars rather than the same sample which often does not fit. But I'm not a sound expert, not sure if I would be able to create proper non-dopplered sample.

\</offtopic\>

**Comment by Moburma** [comment]
Here's a couple of videos of it. Sitting there and actually looking at it properly, it's clearly just the panning extremes of the sound are reversed; left is right and right is left. The second video below makes that obvious:

https://youtu.be/aTQHkHOzdms

https://youtu.be/LNr94uYRMrA

On the sound thing, a guy who contributed to the REMC2 (magic carpet 2) project made proper 3D positional audio for that, it's great, not sure if that code would be useful here.

**Comment by mefistotelis** [comment]
Yeah, that shows the issue well, at least with one specific angle. Makes me wonder whether in perpendicular view the directions would also be reversed.

> On the sound thing, a guy who contributed to the REMC2 (magic carpet 2) project made proper 3D positional audio for that, it's great, not sure if that code would be useful here.

The computations for this are not that complex.
I want the bfsound library to support SDLSound at some point, but OpenAL will likely remain the default.

I was thinking about re-creating the part om MSS library which handles panning - then it would work in the exact same way as original game. But not sure it it's worth the time. Plus, if it's handled by OpenAL it may be made better than original.

Anyway, the code is here, you may experiment with that:
https://github.com/swfans/swars/blob/master/bfsoundlib/src/drv_oal.c#L369

**Comment by Moburma** [comment]
Ok, so I noticed the line that handles panning - the sound is reversed, so let's try inverting the values:

`x_pos = -((127 - s->pan - 64) * (1.f / 64.f));`

That works, but I don't know if it's "technically" a correct thing to do.

Video here:

https://youtu.be/O7joth7cmLc

**Comment by mefistotelis** [comment]
`sample->pan` is value between 0 and 127 , while OpenAL accepts a float between -1.0 and 1.0.
We assume that 1 is full left and 127 is full right (which seem correct because the game sets value 64 to non-3D samples).
Value of 0 - the current equation suggests it's a special value which disables panning?

If that's the case, then to convert pans straight you need `(s->pan - 64)` and to convert reversed you need `(127 - s->pan - 64)`.

From what I understand, channels should be reversed only when `DIG_PCM_ORDER` is set within sample flags.

**Comment by Moburma** [comment]
I see. Assuming it is "straight", works with the following simplification:

` x_pos = (s->pan - 64) * (1.f / 64.f);`

https://youtu.be/XyqBCGThqS4

**Comment by geist22** [comment]
For me this only fixes the x axis (from west to east on the briefing map), but if you rotate the camera 90° so left/right is north/south, it's still inverted.

**Comment by Moburma** [comment]
Urgh, yes you are right. I guess that is why it was set how it was in the first place - possibly at that angle it needs to be inverted? 
I wonder if there is a view angle or something that can be exported from e.g. ASM_do_rotate_map that could be used to say "If X to Y degrees view invert, else don't"?

**Comment by mefistotelis** [comment]
I would prefer a proper solution rather than a hack.

I guess it will have to wait for proper solving.
